demo/demo/tests/test_functions.py

```diff
@@ -215,6 +215,49 @@ def test_multiple_subscriptions(db, two_subscriptions, user, resource, now, rema
     assert remains(at=now + days(16)) == 0
 
 
+def test_multiple_subscriptions_refreshes(db, two_subscriptions, user, resource, now, refreshes, days):
+    pairs = [
+        (sub.start, Quota.objects.get(plan=sub.plan).recharge_period, sub.end)
+        for sub in two_subscriptions
+    ]
+
+    def assert_expected(at: datetime) -> None:
+        output = []
+
+        for sub_start, recharge_period, sub_end in pairs:
+            if sub_start > at:
+                continue
+
+            for idx in count(start=0):
+                moment = sub_start + idx * recharge_period
+                if moment >= sub_end:
+                    break
+                if moment >= at:
+                    output.append(moment)
+                    break
+
+        if not output:
+            expected_result = None
+        else:
+            expected_result = min(output)
+
+        assert refreshes(at=at, assume_subscription_refresh=False) == expected_result, output
+
+    assert_expected(at=now - days(1))
+    assert_expected(at=now + days(0))
+    assert_expected(at=now + days(1))
+    assert_expected(at=now + days(2))
+    assert_expected(at=now + days(4))
+    assert_expected(at=now + days(5))
+    assert_expected(at=now + days(6))
+    assert_expected(at=now + days(7))
+    assert_expected(at=now + days(9))
+    assert_expected(at=now + days(10))
+    assert_expected(at=now + days(11))
+    assert_expected(at=now + days(12))
+    assert_expected(at=now + days(16))
```

Comment:
I'm not sure here, so it's not a request to change these lines; but fyi I would just hardcode the numbers, cause 1) you may make mistakes in your testing algo, and 2) you compare "what you coded" (for test) to "what you coded" (for production), which may fail if you make the same error in both versions - since you're the same person and you're writing the same logic in both cases.

Not about this particular PR, but in general. I'm curious wdyt

Reply:
That's overall a good idea, it just makes the tests more rigid. But true – it's just code vs code and you don't know then which of them contains a mistake. I'll make it "better", I admit I was being lazy here.